File: raspberry/SmartHome/app/src/main/java/raspberry/smarthome/model/GUID.java
package raspberry.smarthome.model;

import java.util.HashSet;
import java.util.Set;

import raspberry.smarthome.model.device.IotDevice;
import raspberry.smarthome.model.device.controllers.ArduinoController;
import raspberry.smarthome.model.device.controllers.BaseController;

/**
 * IoT device can be identified by its guid. As well as its controllers
 */
public class GUID {
    private static GUID instance;
    private final Set<Long> mGuids = new HashSet<>();


    private GUID() {}

    public static GUID getInstance() {
        if (instance == null) {
            instance = new GUID();
        }
        return instance;
    }

    public long getGuidForIotDevice(IotDevice device) {
        return getGuidForName(device.name);
    }

    public long getGuidForController(BaseController controller) {
        if (!(controller instanceof ArduinoController)) {
            throw new RuntimeException("Not implemented"); // todo implement for non arduino controller
        }

        ArduinoController aController = (ArduinoController) controller;

        long guid = getGuidForIotDevice(aController.device)
                * aController.type.id
                * (aController.indexInArduinoServicesArray + 1);
        long shift = 1;
        while (mGuids.contains(guid)) {
            guid += shift;
        }
        // so we've created a unique guid for the controller
        return guid;
    }


    private long getGuidForName(String name) {
        return name.hashCode();
    }
}

File: smarthome_libraries/common/src/main/java/smarthome/library/common/GUID.kt
package smarthome.library.common

import java.text.SimpleDateFormat
import java.util.*
import kotlin.random.Random

/**
 * IoT device can be identified by its guid. As well as its controllers
 */
@Deprecated("use Id.kt instead")
object GUID {
    private val mGuids = HashSet<Long>()

    val guidForMessage: String
        get() {
            val sdf = SimpleDateFormat("dd.MM.yyyy_HH:mm:ss")
            return sdf.format(Date())
        }

    fun remove(guid: Long) {
        mGuids.remove(guid)
    }

    fun getGuidForName(name: String): Long {
        return name.hashCode().toLong()
    }

    fun random(): Long {
        return Random.nextLong()
    }
}
